src/logic/user-items-filter.ts

```diff
@@ -0,0 +1,150 @@
+  shouldExtendWearables: boolean
+) {
+  return wearablesUrn
+    .map((wearable: string) => {
+      if (wearable.includes('base-avatars')) {
+        return wearable
+      }
+
+      const { urn, tokenId } = splitUrnAndTokenId(wearable)
+
+      const matchingOwnedWearable = ownedWearables.find(
+        (ownedWearable) =>
+          ownedWearable.urn === urn &&
+          (!tokenId || ownedWearable.individualData.find((itemData) => itemData.tokenId === tokenId))
+      )
+
+      if (!matchingOwnedWearable) {
+        return undefined
+      }
+
+      return shouldExtendWearables
+        ? `${matchingOwnedWearable.urn}:${tokenId ? tokenId : matchingOwnedWearable.individualData[0].tokenId}`
+        : matchingOwnedWearable.urn
+    })
+    .filter((wearable) => !!wearable) as string[]
+}
+
+export async function createUserItemsFilter(
+  components: Pick<BaseComponents, 'config' | 'wearablesFetcher' | 'emotesFetcher'>
+): Promise<UserItemsFilter> {
+  const { wearablesFetcher, emotesFetcher, config } = components
+  // If ERC-721 Standard is enabled, all items will be extended to contain the tokenId at the end of the urn.
+  const ensureERC721 = (await config.getString('ENSURE_ERC_721')) === 'true'
+
+  /**
+   * Filter out wearables and emotes that are not owned by the user.
+   *
+   * @param {Avatar[]} avatars
+   * @param {string} owner
+   * @return {*}  {Promise<Avatar[]>}
+   */
+  async function filterNotOwnedItemsFromAvatars(avatars: Avatar[], owner: string): Promise<Avatar[]> {
+    const ownedWearables: OnChainWearable[] = await wearablesFetcher.fetchOwnedElements(owner)
+    const ownedEmotes: OnChainEmote[] = await emotesFetcher.fetchOwnedElements(owner)
+
+    // Filter out wearables and emotes that are not owned by the user.
+    const sanitizedAvatars: Avatar[] = avatars.map((avatar: Avatar) => {
+      const validatedWearables: string[] = parseValidWearablesAndFilterInvalidOnes(
+        avatar.avatar.wearables,
+        ownedWearables,
+        ensureERC721
+      )
+
+      const validatedEmotes = avatar.avatar.emotes
+        ?.map((emote: { urn: string; slot: number }) => {
+          if (!emote.urn.includes(':')) {
+            return emote
+          }
+
+          let urnToReturn: string = emote.urn
+
+          const { urn, tokenId } = splitUrnAndTokenId(emote.urn)
+
+          const matchingOwnedEmote = ownedEmotes.find(
+            (ownedEmote) =>
+              ownedEmote.urn === urn &&
+              (!tokenId || ownedEmote.individualData.find((itemData) => itemData.tokenId === tokenId))
+          )
+
+          if (!matchingOwnedEmote) {
+            return undefined
+          }
+
+          urnToReturn = ensureERC721
+            ? `${matchingOwnedEmote.urn}:${tokenId ? tokenId : matchingOwnedEmote.individualData[0].tokenId}`
+            : matchingOwnedEmote.urn
+
+          return { urn: urnToReturn, slot: emote.slot }
+        })
+        .filter((emote) => !!emote) as { urn: string; slot: number }[]
```

Comment:
I would try to avoid the `as ..` it's very error-prone, I think in this case you could rewrite this as a for loop, I know it's a bit uglier but you would avoid typescript incorrectly infering types